### PR TITLE
Fine-grained access control for API Gateway using Lambda Authorizer

### PR DESCRIPTION
**Pull request by VikramPatil-AMZ**
Fine-grained access control for API Gateway using Lambda Authorizer. This C# CDK pattern creates Amazon Cognito user pool, Amazon API Gateway, AWS Lambda auth function, Amazon DynamoDB table, and AWS Lambda backend function. 

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

**Review by Talalgmel — CHANGES_REQUESTED** [review]
Talal

**Comment by pahud** [comment]
Hi, thanks for your PR! I will be happy to verify the deployment in my environment to help your PR better get approved and merged.

I was following your steps and when I run this command in the `apigateway-cognito-lambda-dynamodb` directory:

`dotnet publish src\Lambda\BackendFunction\BackendFunction.csproj -c Release -o dist/BackendFunction`

I got the `Project file does not exist` error.

```
Microsoft (R) Build Engine version 16.7.3+2f374e28e for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

MSBUILD : error MSB1009: Project file does not exist.
Switch: srcLambdaBackendFunctionBackendFunction.csproj
```

$ dotnet --version
3.1.424

Did I miss anything?

**Comment by VikramPatil-AMZ** [comment]
> Hi, thanks for your PR! I will be happy to verify the deployment in my environment to help your PR better get approved and merged.
> 
> I was following your steps and when I run this command in the `apigateway-cognito-lambda-dynamodb` directory:
> 
> `dotnet publish src\Lambda\BackendFunction\BackendFunction.csproj -c Release -o dist/BackendFunction`
> 
> I got the `Project file does not exist` error.
> 
> ```
> Microsoft (R) Build Engine version 16.7.3+2f374e28e for .NET
> Copyright (C) Microsoft Corporation. All rights reserved.
> 
> MSBUILD : error MSB1009: Project file does not exist.
> Switch: srcLambdaBackendFunctionBackendFunction.csproj
> ```
> 
> $ dotnet --version 3.1.424
> 
> Did I miss anything?

Hi Pahud, 

Updated the Readme file to replace backward slash with forward slash, so that will not get this error on non-windows terminal (forward slash works with windows terminal as well). 

Thanks for letting me know.

**Comment by indrora** [comment]
(Sorry for the late response -- I thought I sent this earlier)

The README here needs refinement. I'll put in some comments and suggestions for clearer wording.

**Comment by pahud** [comment]
I have successfully deployed this in my AWS Account and verified all steps provided with no outstanding issues except some notes. I think as a CDK sample it's ok to go.

Some notes:

1. When sending the POST request, the API is actually returning `201` rather than `200`. 
2. Postman might not be allowed to install due to license restriction in some environment, consider to offer other alternatives as well such as `Yet Another Rest Client` chrome extension.
3. The `access_token` extraction from the URL location bar was a pain for me. I ended up writing a `sed` to extract it from the provided string. Given some Windows user might not have sed installed, would be great if we provide both `sed` and `python` samples to help extract the `access_token` locally from the given string. Not mandatory, but great to improve user experience.
4. The cold start of the Lambda function in this sample took up to 12s. Not sure if this was the hard limitation of the lambda csharp runtime. Worth a note in the README to avoid too much user surprise. Not a big deal as a sample though.

Overall, it's a great csharp sample. I'm giving a LGTM here. Thanks for the PR contribution.

File: csharp/apigateway-cognito-lambda-dynamodb/README.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+14. Confirm API GW returned 200 Success response
+
+15. Invoke POST endpoint with access token of second user
+
+16. Confirm API GW returned 201 Success response
+
+17. Invoke GET or POST endpoint with invalid token
+
+18. Confirm API GW returned 401 Unauthorized response
```
Comment (indrora) [inline csharp/apigateway-cognito-lambda-dynamodb/README.md:119]:
```suggestion
11. Confirm that you are returned an HTTP 403 Unauthorized response.

12. Log into the Hosted UI as the second user you created following the steps previously described. 

13. Make a GET request to the API Gateway HTTP Endpoint as you did previously, this time using the second user's `access_token` for the bearer token.

14. Confirm that you are returned a 200 Success response

15. Make a POST request to the API HTTP Endpoint 

16. Confirm you are returned an HTTP 201 Created response

17. Try making a GET or POST request to the endpoint with invalid token

18. Observe that you're returned an HTTP 401 Unauthorized response
```

**Review by indrora — CHANGES_REQUESTED** [review]
README needs some work.

File: csharp/apigateway-cognito-lambda-dynamodb/README.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+> **This is a stable example. It should successfully build out of the box**
+>
+> This examples is built on Construct Libraries marked "Stable" and does not have any infrastructure
+> prerequisites to build.
+
+---
+
+<!--END STABILITY BANNER-->
+
+This C# CDK pattern creates Amazon Cognito user pool, Amazon API Gateway, AWS Lambda auth function, Amazon DynamoDB table, and AWS Lambda backend function. Auth Lambda function verify JWT token received in request header returns access policy associated with token user group. Based on access policy, API GW decides to forward the request to backend Lambda function or return 401/403 http status.
```
[the remark this answers is not in the document]
Comment (VikramPatil-AMZ) [reply csharp/apigateway-cognito-lambda-dynamodb/README.md]: awesome suggestions. Updated readme file.

**Comment by VikramPatil-AMZ** [comment]
> I have successfully deployed this in my AWS Account and verified all steps provided with no outstanding issues except some notes. I think as a CDK sample it's ok to go.
> 
> Some notes:
> 
> 1. When sending the POST request, the API is actually returning `201` rather than `200`.
> 2. Postman might not be allowed to install due to license restriction in some environment, consider to offer other alternatives as well such as `Yet Another Rest Client` chrome extension.
> 3. The `access_token` extraction from the URL location bar was a pain for me. I ended up writing a `sed` to extract it from the provided string. Given some Windows users might not have sed installed, would be great if we provide both `sed` and `python` samples to help extract the `access_token` locally from the given string. Not mandatory, but great to improve user experience.
> 4. The cold start of the Lambda function in this sample took up to 12s. Not sure if this was the hard limitation of the lambda csharp runtime. Worth a note in the README to avoid too much user surprise. Not a big deal as a sample though.
> 5. We tend to provide AWS CLI commands over UI to avoid potential UI deprecation. Would be great if we can provide AWS CLI commands to create user pool users and add them into the groups.
> 6. Love the final clean up with AWS CLI verification. I've learned as well.
> 
> Overall, it's a great csharp sample. I'm giving a LGTM here. Thanks for the PR contribution.

Hi Pahud, 
Thanks for the review. 

Updated read me file for comment#1 - POST endpoint response returns 201, comment#2 - Added comment for REST client along with Postman, 
For remaining comments - 
comment#3, and comment#5 - I will keep checking on AWS CLI commands as follow up improvement for this implementation.  
comment#4 - increased Lambda memory to 216MB, and I can see 3secs-4secs cold start at my end, I will keep checking on performance improvement as follow up improvement.

**Review by indrora — APPROVED** [review]

**Comment by indrora** [comment]
A follow-up PR can address most of these issues. Unfortunately, registering users through the AWS CLI can be a bit convoluted and is easier done through the UI.

I agree that a cleaner approach to extracting the access_token parameter would be useful. Unfortunately, there's no good mix-in solution at the moment. I'd like to see the Cognito team offer a "debug endpoint" that you can redirect to that shows what is passed to it. Using `sed` and `python` is a non-starter due to sed being nonexistent in Windows without additional third party software (e.g. Cygwin, mSYS2) or WSL2. 

.NET cold start is a known issue for the Lambda team and where there is room to improve mostly on the .NET side; This has been approached by a variety of different groups in the past, with some of them achieving some impressive numbers:

* https://medium.com/slalom-build/solving-cold-starts-on-aws-lambda-when-using-dotnet-core-51f244f08f60
* https://blog.thundra.io/solving-dotnet-lambda-cold-start-part-1
* https://zaccharles.medium.com/making-net-aws-lambda-functions-start-10x-faster-using-lambdanative-8e53d6f12c9c

This has also been brought up to the .NET Core team: https://github.com/dotnet/core/issues/1060

A faster alternative is Go or NodeJS, both of which have fast startup times from cold start. 

As for Postman, it's an unfortunate baseline. Providing a cURL command line is a close second.

**Pull request merged**